### DISCUSSION TITLE
[Guide] Debugging GitLens in an Extension Host Process After It’s Started

### DISCUSSION THREAD

**Discussion by gluxon** [gitkraken/vscode-gitlens#3068] [Feedback]
## Context

I've been running into a strange GitLens issue (https://github.com/gitkraken/vscode-gitlens/issues/3066) for a long time, but had trouble nailing it down since the problem was always intermittent and I wasn't able to reproduce it in a [Extension Development Window](https://code.visualstudio.com/api/get-started/your-first-extension). When the problem did reproduce, it went away with a extension host restart, removing the repro.

## Goal

It should be possible to take any production built VS Code window running an extension like GitLens and debug it without a restart that may lose the bad state. I recently worked through how to do this and wanted to write it down somewhere for others (and my future self).

## Disclaimers

- This is a write-up from an external contributor. I'm not affiliated with GitLens.
- This will likely be out of date in a few years.
- There will probably be VS Code or Chrome improvements that make this easier in the future.

### Finding the VS Code Extension Host Process Port

The usual trick for debugging a `node` process is to [send it the `SIGUSR1` signal](https://nodejs.org/en/guides/debugging-getting-started#enable-inspector).

```
❯ kill -SIGUSR1 <pid>
```

The VS Code extension host process seems to listen on a less deterministic port than `node` processes out of the box. Fortunately VS Code makes it easy to attach itself to its own extension host.

<img width="1136" alt="Screenshot 2023-12-20 at 2 02 20 AM" src="https://github.com/gitkraken/vscode-gitlens/assets/906558/df9b5c58-635c-49b3-8456-980abf5e88c5">

Hit the circle button to start profiling the extension host. Then hit the play button to the left to attach to the process. (Without hitting the profile button first, VS Code offers to restart the extension host.)

<img width="1136" alt="Screenshot 2023-12-20 at 2 02 50 AM" src="https://github.com/gitkraken/vscode-gitlens/assets/906558/587b9a22-bb41-46a3-949d-23c96bf5984c">

To attach Chrome's Node.js inspector, we'll need the port the process is listening on. In the debug console, we can enter:

```js
require('inspector').url()
```

<img width="1136" alt="Screenshot 2023-12-20 at 2 04 44 AM" src="https://github.com/gitkraken/vscode-gitlens/assets/906558/bd794924-753c-42e0-9325-738dcd1b4c50">

The port for this process is `56929`. 

### Attaching Chrome

At `chrome://inspect`, we'll add the port discovered above.

<img width="1286" alt="Screenshot 2023-12-20 at 2 05 19 AM" src="https://github.com/gitkraken/vscode-gitlens/assets/906558/6be43460-e576-47ec-a746-c83e7ed91329">

This gives a new _Inspect_ button.

<img width="1286" alt="Screenshot 2023-12-20 at 2 05 25 AM" src="https://github.com/gitkraken/vscode-gitlens/assets/906558/0a02bcde-a577-48af-91b3-fa1c4b464074">

From here, we can search for files loaded into the process. For GitLens, one file of interest is the bundled `gitlens.js` file.

<img width="1290" alt="Screenshot 2023-12-20 at 2 05 34 AM" src="https://github.com/gitkraken/vscode-gitlens/assets/906558/d4452bb5-a43a-493b-8ad9-40ca845dc57c">

### Checking Out Source

The above allows us to interactively debug `gitlens.js`, but this file is hard to understand since it's bundled and minified. We can GitLen's source code to make this easier. To do so, we'll want to check out the source for the exact version of GitLens we're using.

<img width="1375" alt="Screenshot 2023-12-20 at 1 55 26 AM" src="https://github.com/gitkraken/vscode-gitlens/assets/906558/8fbf7259-4bb6-4757-928e-81167811d175">

```
~/Developer/vscode-gitlens main 
❯ git checkout v14.6.1        
Note: switching to 'v14.6.1'.
```

Let's make a small tweak to GitLen's `webpack.config.js` to create a production build with source maps. We'll want to use `hidden-source-map` to output our own `gitlens.js` file that matches the currently running one.

```diff
diff --git a/webpack.config.js b/webpack.config.js
index 84fd17e9b..097d3a4f9 100644
--- a/webpack.config.js
+++ b/webpack.config.js
@@ -150,7 +150,7 @@ function getExtensionConfig(target, mode, env) {
                },
                mode: mode,
                target: target,
-               devtool: mode === 'production' ? false : 'source-map',
+               devtool: 'hidden-source-map',
                output: {
                        chunkFilename: 'feature-[name].js',
                        filename: 'gitlens.js',
```

To build GitLens for production:

```
❯ yarn bundle
```

As an optional sanity check, we can make sure the built `gitlens.js` files are the same as what VS Code is running. If there are mismatches, the source code cloned could be subtly different than the original source of the loaded `gitlens.js` file, which would cause breakpoints to mismatch and other oddities when debugging.

```
❯ shasum ~/.vscode/extensions/eamodio.gitlens-14.6.1/dist/gitlens.js dist/gitlens.js
813627ea9dc246815b1077bed4ab4c50eda228a5  /Users/gluxon/.vscode/extensions/eamodio.gitlens-14.6.1/dist/gitlens.js
813627ea9dc246815b1077bed4ab4c50eda228a5  dist/gitlens.js
```

### Adding Source Maps to Chrome's Inspector

Let's teach Chrome about our built `dist/gitlens.js.map`. We'll need the absolute path of this file.

<img width="1136" alt="Screenshot 2023-12-20 at 2 06 32 AM" src="https://github.com/gitkraken/vscode-gitlens/assets/906558/0d9b8ee9-ddfd-433e-b56d-704f77da3659">

Back in the Chrome inspector, we can right click anywhere in `gitlens.js` to "_Add source map..._"

<img width="1290" alt="Screenshot 2023-12-20 at 2 45 27 AM" src="https://github.com/gitkraken/vscode-gitlens/assets/906558/2d01ca2f-60b8-484e-8565-769de9099a9a">

<img width="1290" alt="Screenshot 2023-12-20 at 2 06 47 AM" src="https://github.com/gitkraken/vscode-gitlens/assets/906558/cd14eff8-dcd8-43f3-88f8-77abbe559293">

Finally, let's add our cloned `vscode-gitlens` directory to make file navigation easier. Hit "_Add folder_" on the left sidebar.

<img width="1290" alt="Screenshot 2023-12-20 at 2 07 05 AM" src="https://github.com/gitkraken/vscode-gitlens/assets/906558/231e5a4f-a582-4d81-a76f-02826dd1e646">

### Profit

It works! Note that the files loaded into Chrome on the left have green circles indicating their source maps are correctly associated with the generated `gitlens.js` production build.

<img width="1290" alt="Screenshot 2023-12-20 at 2 08 26 AM" src="https://github.com/gitkraken/vscode-gitlens/assets/906558/132cd497-19c5-40cd-9a51-6e02455911e5">

**Comment by eamodio** [comment]
@gluxon This is awesome -- TY!

**Reply by gluxon** [reply]
Thank you! Appreciate all your work on GitLens. Love it.
